### PR TITLE
fix(useRouteQuery) Set undefined when null, undefined or default value

### LINKED ISSUE #3290: Unset route query on default value is not working anymore
## Setup

Package: @vueuse/router

```js
<script setup>
import { useRouteQuery } from "@vueuse/router";
const search = useRouteQuery("search", "defaultValue");
</script>
```

## Expected Beahviour and Behaviour of Version v10.2.1

When I change the value of `search` to the default value (in this case `"defaultValue"`), the query parameter is removed from the URL:
```js
search.value = "defaultValue";
```

## Behaviour in Version  v10.3.0

When changing the value of `search` to the default value (in this case `"defaultValue"`), the parameter remains in the URL.

## Further Notes
In my production use case the default value is `""` which leads to to a paramter `?search=`

## Further observations

* It was intended behaviour in the past, as far as I can tell: https://github.com/vueuse/vueuse/pull/1490
* It was remove in https://github.com/vueuse/vueuse/commit/0d02f6cc6a0a78374ab81e13f594f16eb4695b6a as part of https://github.com/vueuse/vueuse/pull/3184
* There is one Discussion about the same method, but is more about a changed behaviour https://github.com/vueuse/vueuse/discussions/2914

### Reproduction

https://github.com/greenhalos/useRouteQuery-bug-default

### System Info

```Shell
Binaries:
    Node: 18.15.0 - ~/.nvm/versions/node/v18.15.0/bin/node
    npm: 9.5.0 - ~/.nvm/versions/node/v18.15.0/bin/npm
    pnpm: 8.6.11 - ~/.local/share/pnpm/pnpm
  npmPackages:
    @vueuse/core: 10.3.0 => 10.3.0 
    @vueuse/router: 10.3.0 => 10.3.0 
    vue: ^3.2.25 => 3.3.4
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vueuse/vueuse/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vueuse.org/guide).
- [X] Check that there isn't [already an issue](https://github.com/vueuse/vueuse/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a VueUse issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/core instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vueuse/vueuse/discussions).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by greenhalos**
### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vueuse/vueuse/blob/main/packages/guidelines.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Ideally, include relevant tests that fail without this PR but pass with it.

<details>
<summary><strong>⚠️ Slowing down new functions</strong></summary>
<br>

> **Warning**: **Slowing down new functions**
>
> As the VueUse audience continues to grow, we have been inundated with an overwhelming number of feature requests and pull requests. As a result, maintaining the project has become increasingly challenging and has stretched our capacity to its limits. As such, in the near future, we may need to slow down our acceptance of new features and prioritize the stability and quality of existing functions. **Please note that new features for VueUse may not be accepted at this time.** If you have any new ideas, we suggest that you first incorporate them into your own codebase, iterate on them to suit your needs, and assess their generalizability. If you strongly believe that your ideas are beneficial to the community, you may submit a pull request along with your use cases, and we would be happy to review and discuss them. Thank you for your understanding.

</details>

---

### Description

fix #3290 

### Additional context

Also tried to resolve the same issue for useRouteParams

---

### <samp>🤖 Generated by Copilot at 2473669</samp>

This pull request fixes bugs and improves the behavior of `useRouteQuery` and `useRouteParams` functions, which sync refs with route query and param values. It makes them remove the query or param from the route when the ref value is the default, null, or undefined.

### <samp>🤖 Generated by Copilot at 2473669</samp>

*  Remove param or query from route when it matches default value or is null ([link](https://github.com/vueuse/vueuse/pull/3291/files?diff=unified&w=0#diff-f657fa605a6d4b2a9769faa5835fde4154c39c4a01a09cf8ae63e9dac446fcd8L67-R67), [link](https://github.com/vueuse/vueuse/pull/3291/files?diff=unified&w=0#diff-3a643af876b64ca6e9d5db979bfc2cfb0af8d46845a7cd6bfe4b1e41e5a369b5L66-R66))
* Fix tests for `useRouteParams` and `useRouteQuery` to expect undefined instead of 1 when ref is set to default value ([link](https://github.com/vueuse/vueuse/pull/3291/files?diff=unified&w=0#diff-a57f86ebcafcb55d44126c69a48bd1f5c88e513a80fddd3241ecd3ae34ba76ddL174-R174), [link](https://github.com/vueuse/vueuse/pull/3291/files?diff=unified&w=0#diff-faa7aba85ff895209572d55a507c772d3fbc8ff9530641d792cd69d5eeb0101aL195-R195))
* Add test for `useRouteQuery` to check that it resets the value when it is set to default value, null, or undefined ([link](https://github.com/vueuse/vueuse/pull/3291/files?diff=unified&w=0#diff-faa7aba85ff895209572d55a507c772d3fbc8ff9530641d792cd69d5eeb0101aR236-R253))

### CODE REVIEW

File: packages/router/useRouteParams/index.ts [hunk 1]
```diff
@@ -64,7 +64,7 @@ export function useRouteParams<
         if (_params.get(name) === v)
           return
 
-        _params.set(name, v)
+        _params.set(name, (v === defaultValue || v === null) ? undefined : v)
```
Comment (antfu) [inline packages/router/useRouteParams/index.ts:67]:
```suggestion
        _params.set(name, v == null ? undefined : v)
```

File: packages/router/useRouteQuery/index.ts [hunk 2]
```diff
@@ -63,7 +63,7 @@ export function useRouteQuery<
         if (_query.get(name) === v)
           return
 
-        _query.set(name, v)
+        _query.set(name, (v === defaultValue || v === null) ? undefined : v)
```
Comment (antfu) [inline packages/router/useRouteQuery/index.ts:66]:
```suggestion
        _query.set(name, v == null ? undefined : v)
```

**Comment by antfu** [comment]
I think this should be considered as a breaking change.

/cc @anteriovieira What's your take on if we should move forward with this?

File: packages/router/useRouteParams/index.ts (hunk 1 above)
Comment (greenhalos) [reply packages/router/useRouteParams/index.ts:67]: But this breaks it again in case it is the `defaultValue` or did you just meant to change the last part `=== null` and leave the first part as is?

```suggestion
        _params.set(name, (v === defaultValue || v == null) ? undefined : v)
```

Comment (greenhalos) [reply packages/router/useRouteParams/index.ts:67]: At least my added test cases break

**Comment by greenhalos** [comment]
I have rebased to the latest commits on the main branch. Could you please review again? @antfu @anteriovieira
